vm/systemSmartContracts/esdt.go

```diff
@@ -795,6 +805,206 @@ func (e *esdt) controlChanges(args *vmcommon.ContractCallInput) vmcommon.ReturnC
 	return vmcommon.Ok
 }
 
+func (e *esdt) checkArgumentsForSpecialRoleChanges(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if len(args.Arguments) < 3 {
+		e.eei.AddReturnMessage("not enough arguments")
+		return vmcommon.FunctionWrongSignature
+	}
+	if len(args.Arguments[1]) != len(args.CallerAddr) {
+		e.eei.AddReturnMessage("first argument must be an address")
+		return vmcommon.FunctionWrongSignature
+	}
+	err := validateRoles(args.Arguments[2:])
+	if err != nil {
+		e.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	return vmcommon.Ok
+}
+
+func (e *esdt) setSpecialRole(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	returnCode := e.checkArgumentsForSpecialRoleChanges(args)
+	if returnCode != vmcommon.Ok {
+		return returnCode
+	}
+
+	token, returnCode := e.basicOwnershipChecks(args)
+	if returnCode != vmcommon.Ok {
+		return returnCode
+	}
+
+	address := args.Arguments[1]
+	esdtRole, isNew := getRolesForAddress(token, address)
+	if isNew {
+		esdtRole.Roles = append(esdtRole.Roles, args.Arguments[2:]...)
+		token.SpecialRoles = append(token.SpecialRoles, esdtRole)
+	} else {
+		for _, arg := range args.Arguments[2:] {
+			_, exist := doesRoleExist(esdtRole, arg)
+			if !exist {
+				esdtRole.Roles = append(esdtRole.Roles, arg)
+			}
+		}
+	}
+
+	err := e.sendRoleChangeData(args, core.BuiltInFunctionSetESDTRole)
+	if err != nil {
+		e.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	err = e.saveToken(args.Arguments[0], token)
+	if err != nil {
+		e.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	return vmcommon.Ok
+}
+
+func (e *esdt) unSetSpecialRole(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	returnCode := e.checkArgumentsForSpecialRoleChanges(args)
+	if returnCode != vmcommon.Ok {
+		return returnCode
+	}
+	token, returnCode := e.basicOwnershipChecks(args)
+	if returnCode != vmcommon.Ok {
+		return returnCode
+	}
+
+	address := args.Arguments[1]
+	esdtRole, isNew := getRolesForAddress(token, address)
+	if isNew {
+		e.eei.AddReturnMessage("address does not have special role")
+		return vmcommon.UserError
+	}
+
+	for _, arg := range args.Arguments[2:] {
+		index, exist := doesRoleExist(esdtRole, arg)
+		if !exist {
+			e.eei.AddReturnMessage("role does not exist for address")
+			return vmcommon.UserError
+		}
+
+		copy(esdtRole.Roles[index:], esdtRole.Roles[index+1:])
```

Comment:
Duplicated code with esdtRoles.go L84-L86.

Reply:
yeah. but they are in completely different packages. with different data types as well.